### PR TITLE
feature: tem refactoring + tutorial notebook

### PR DESCRIPTION
**Pull request by jedyeo**

### CODE REVIEW

**Comment by geoffwoollard** [comment]
This looks incomplete. I put some comments in https://app.reviewnb.com/compSPI/simSPI/pull/86/

File: tests/test_notebook.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        "-p",
+        "sim_config",
+        sim_yaml,
+        "--execution-timeout",
+        "180",
+    ]
+    subprocess.check_call(args)
+
+
+def test_tem_tutorial(test_resources):
```
Comment (ninamiolane) [inline tests/test_notebook.py:46]: Nice!

File: tests/test_tem_inputs.py [hunk 2]
```diff
@@ -436,13 +465,130 @@ def test_write_inp_file():
...
         os.unlink(tmp_yml.name)
+
+
+def test_write_tem_defocus_file_from_distribution(tmp_path):
+    """Test if defocus file is generated with right format."""
+    test_defocus_file = str(Path(tmp_path, "defocus_file_test.txt"))
+    test_distribution_len = 10
+    test_distribution = list(np.random.rand(test_distribution_len))
+
+    print(test_distribution)
```
Comment (ninamiolane) [inline tests/test_tem_inputs.py:487]: left-over prints?

**Review by ninamiolane — APPROVED** [review]
This PR is too long for me to review it properly, but I am OK to merge if someone else can have a look and approve it.

**Comment by fredericpoitevin** [comment]
Is this ready to review? I see some tests not passing

**Comment by arjunsingh3600** [comment]
> Is this ready to review? I see some tests not passing
I merged this branch to reflect changes by #82. Since some of the changes re-introduced changed some file paths - the tests will have to be run without cachce. I have marked the branch as a draft while the tests are rerun.

**Comment by geoffwoollard** [comment]
@arjunsingh3600 I merged https://github.com/compSPI/simSPI/pull/82

File: notebooks/tem_tutorial.ipynb [hunk 3]
```diff
@@ -0,0 +1,457 @@
+{
```
Comment (geoffwoollard) [inline notebooks/tem_tutorial.ipynb:1]: <p>Ensure the outputs to the notebook are available/run in the committed version that is publicly available.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/compSPI/simSPI/pull/87/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (geoffwoollard) [inline notebooks/tem_tutorial.ipynb:1]: <p>image not showing up</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/compSPI/simSPI/pull/87/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

File: tests/test_files/tem/sim_config.yaml [hunk 4]
```diff
@@ -19,16 +21,16 @@
...
-    spherical_aberration_mm: 2.7     # the spherical aberration of the optical system in mm.
-    chromatic_aberration_mm: 2.7     # the chromatic aberration of the optical system in mm.
-    aperture_diameter_um: 50         # diameter of aperture in back focal plane in μm.
-    focal_length_mm: 3.5             # the focal length of primary lens in mm.
-    aperture_angle_mrad: 0.1         # the aperture angle of beam furnished by condenser lens in mrad.
-    defocus_um: 1.0                  # [μm]
-    defocus_syst_error_um: 0.0       # [μm]
-    defocus_nonsyst_error_um: 0.0    # [μm]
-    optics_defocusout: None     # file to write defocus values
+    magnification: 81000             #
```
Comment (ninamiolane) [inline tests/test_files/tem/sim_config.yaml:24]: Why removing the descriptions of the parameters?

**Review by ninamiolane — APPROVED** [review]
Thanks! LGTM, just one comment about why the descriptions of some parameters have been removed.

File: tests/test_files/tem/sim_config.yaml (hunk 4 above)
Comment (arjunsingh3600) [reply tests/test_files/tem/sim_config.yaml]: Thank you for catching this. These seem to have been accidentally removed during a merge conflict. The descriptions have been added back.

**Comment by geoffwoollard** [comment]
Hooray!!! 

Look at that beautiful test coverage!

Look back and see how far you've come, enjoy, and imagine how this could come together in the time you have left. 

Keep working hard!

```
---------- coverage: platform linux, python 3.9.12-final-0 -----------
Name                                          Stmts   Miss  Cover   Missing
---------------------------------------------------------------------------
simSPI/__init__.py                                1      0   100%
simSPI/crd.py                                    27      0   100%
simSPI/fov.py                                    76      0   100%
simSPI/linear_simulator/ctf.py                   37      0   100%
simSPI/linear_simulator/linear_simulator.py      33      0   100%
simSPI/linear_simulator/noise_utils.py            8      0   100%
simSPI/linear_simulator/params_utils.py          87      0   100%
simSPI/linear_simulator/projector.py             20      0   100%
simSPI/linear_simulator/shift_utils.py           22      0   100%
simSPI/linear_simulator/volume_utils.py           7      0   100%
simSPI/multislice.py                             31      0   100%
simSPI/tem.py                                   122      5    96%   192-193, 196, 242-246
simSPI/tem_distribution.py                       18      0   100%
simSPI/tem_inputs.py                            203      6    97%   354-355, 384-385, 504-505
simSPI/transfer.py                               52      3    94%   100, 107, [177](https://github.com/compSPI/simSPI/runs/5685549113?check_suite_focus=true#step:4:177)
---------------------------------------------------------------------------
TOTAL                                           744     14    98%
Coverage XML written to file coverage.xml

Required test coverage of 90.0% reached. Total coverage: 98.12%
================= 57 passed, 28 warnings in 637.83s (0:10:37) ==================
```

**Review by geoffwoollard — APPROVED** [review]

**Review by fredericpoitevin — APPROVED** [review]
Good work, guys!

**Pull request merged**